### PR TITLE
feat(sampling): Implement denylist query param in project tags API [TET-113]

### PR DESCRIPTION
**Pull request by andriisoldatenko**
Add new optional query param `onlySamplingTags` to
`/api/0/projects/<organization>/<project>/tags/`.
So we can filter out tags using sentry.constants.DS_DENYLIST.

Fixes TET-113

### CODE REVIEW

File: src/sentry/api/endpoints/project_tags.py [hunk 1]
```diff
@@ -15,15 +15,19 @@ def get(self, request: Request, project) -> Response:
         except Environment.DoesNotExist:
             tag_keys = []
         else:
+            kwargs = dict(
+                # We might be able to stop including these values, but this
+                # is a pretty old endpoint, so concerned about breaking
+                # existing api consumers.
+                include_values_seen=True,
+            )
+            only_sampling_tags = request.GET.get("onlySamplingTags") == "1"
```
Comment (ahmedetefy) [inline src/sentry/api/endpoints/project_tags.py:24]:
```suggestion
```

File: src/sentry/api/endpoints/project_tags.py [hunk 2]
```diff
@@ -15,15 +15,19 @@ def get(self, request: Request, project) -> Response:
...
+
+            if only_sampling_tags:
```
Comment (ahmedetefy) [inline src/sentry/api/endpoints/project_tags.py:26]:
```suggestion
            if request.GET.get("onlySamplingTags") == "1":
```

File: src/sentry/constants.py [hunk 3]
```diff
@@ -580,3 +580,25 @@ def from_str(cls, string: str) -> Optional[int]:
...
+        "browser",
+        "browser.name",
+        "device",
+        "device.family",
+        "environment",
+        "gpu.name",
+        "gpu.vendor",
+        "level",
+        "logger",
+        "monitor.id" "os",
```
Comment (andriisoldatenko) [inline src/sentry/constants.py:597]: typo, on it.

File: tests/snuba/api/endpoints/test_project_tags.py [hunk 4]
```diff
@@ -31,3 +31,27 @@ def test_simple(self):
         assert data["foo"]["uniqueValues"] == 1
         assert data["bar"]["canDelete"]
         assert data["bar"]["uniqueValues"] == 2
+
+    def test_simple_with_deny(self):
```
Comment (ahmedetefy) [inline tests/snuba/api/endpoints/test_project_tags.py:35]:
```suggestion
    def test_simple_remove_tags_in_denylist(self):
```

File: tests/snuba/tagstore/test_tagstore_backend.py [hunk 5]
```diff
@@ -212,6 +212,31 @@ def test_get_tag_keys(self):
...
+            )
+        }
+        assert set(keys) == expected_keys
+        keys = {
+            k.key: k
+            for k in self.ts.get_tag_keys(
+                project_id=self.proj1.id, environment_id=self.proj1env1.id, denylist=denylist_keys
+            )
+        }
+        assert set(keys) == expected_keys
```
Comment (ahmedetefy) [inline tests/snuba/tagstore/test_tagstore_backend.py:238]: What is the point of this portion of the test?

File: tests/snuba/tagstore/test_tagstore_backend.py [hunk 6]
```diff
@@ -212,6 +212,31 @@ def test_get_tag_keys(self):
         }
         assert set(keys) == expected_keys
 
+    def test_get_tag_keys_with_denylist(self):
```
Comment (ahmedetefy) [inline tests/snuba/tagstore/test_tagstore_backend.py:215]:
```suggestion
    def test_get_tag_keys_removed_from_denylist(self):
```

**Review by ahmedetefy — CHANGES_REQUESTED** [review]

File: tests/snuba/tagstore/test_tagstore_backend.py (hunk 5 above)
Comment (andriisoldatenko) [reply tests/snuba/tagstore/test_tagstore_backend.py:238]: to hit 100% code coverage? but i assume another test can do it right?

Comment (andriisoldatenko) [reply tests/snuba/tagstore/test_tagstore_backend.py:238]: yeah i copied it from another test, 100% we can simplify this test.

Comment (ahmedetefy) [reply tests/snuba/tagstore/test_tagstore_backend.py:238]: So I'd recommend testing within this test once with the deny list and once without for completeness

Comment (andriisoldatenko) [reply tests/snuba/tagstore/test_tagstore_backend.py]: fixed in fac7ee5bf8

File: tests/snuba/tagstore/test_tagstore_backend.py (hunk 6 above)
Comment (andriisoldatenko) [reply tests/snuba/tagstore/test_tagstore_backend.py]: fixed in fac7ee5bf8

File: src/sentry/api/endpoints/project_tags.py (hunk 2 above)
Comment (andriisoldatenko) [reply src/sentry/api/endpoints/project_tags.py]: fixed in fac7ee5bf8

File: src/sentry/api/endpoints/project_tags.py (hunk 1 above)
Comment (andriisoldatenko) [reply src/sentry/api/endpoints/project_tags.py]: fixed in fac7ee5bf8

**Comment by ahmedetefy** [comment]
You mind posting links to the modules you got the DENY_LIST from? Will certainly help me review the DENY_LIST

**Comment by andriisoldatenko** [comment]
> You mind posting links to the modules you got the DENY_LIST from? Will certainly help me review the DENY_LIST

yeah currently trying to explain it using words :)

File: tests/snuba/api/endpoints/test_project_tags.py (hunk 4 above)
Comment (andriisoldatenko) [reply tests/snuba/api/endpoints/test_project_tags.py]: fixed in 039bf58204

File: src/sentry/constants.py [hunk 7]
```diff
@@ -580,3 +580,38 @@ def from_str(cls, string: str) -> Optional[int]:
...
+# `_derive_interface_tags_many(jobs)` which iterates across all interfaces
+# and execute `iter_tags`, so i've searched usage of `iter_tags`.
+# 2. `src/sentry/event_manager.py:_pull_out_data` we have `set_tag`.
+# 3. `src/sentry/event_manager.py:_get_event_user_many` we have `set_tag`.
+# 4. `src/sentry/event_manager.py:_get_or_create_release_many` we have `set_tag`.
+#
+# Note:
+# should be sorted alphabetically for easy to maintanin in future
+# if you update this list please add explanation or source of it
+DS_DENYLIST = frozenset(
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply src/sentry/constants.py:598]: fixed in 285284e0d6

**Review by ahmedetefy — APPROVED** [review]

**Review by matejminar — APPROVED** [review]
Tested locally, looks good to me.
Here's a frontend part: https://github.com/getsentry/sentry/pull/35183

**Comment by andriisoldatenko** [comment]
@jan-auer PTAL

**Pull request merged**